src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,18 +7166,88 @@ bool canFollowNullableType()
                     case ParseTypeMode.AfterIs:
                     case ParseTypeMode.DefinitePattern:
                     case ParseTypeMode.AsExpression:
-                        // These contexts might be a type that is at the end of an expression. In these contexts we only
-                        // permit the nullable qualifier if it is followed by a token that could not start an
-                        // expression, because for backward compatibility we want to consider a `?` token as part of the
-                        // `?:` operator if possible.
+                        // We are currently after `?` token after a nullable type pattern and need to decide how to
+                        // parse what we see next.  In the case of an identifier (e.g. `x ? a` there are two ways we can
+                        // see things
                         //
-                        // Similarly, if we have `T?[]` we do want to treat that as an array of nullables (following
-                        // existing parsing), not a conditional that returns a list.
+                        // 1. As a start of conditional expression, e.g. `var a = obj is string ? a : b`
+                        // 2. As a designation of a nullable-typed pattern, e.g. `if (obj is string? str)`
+                        //
+                        // Since nullable types (no matter reference of value types) are not valid in patterns by
+                        // default we are biased towards the first option and consider case 2 only for error recovery
+                        // purposes (if we parse here as nullable type pattern an error will be reported during
+                        // binding). This condition checks for simple cases, where we better use option 2 and parse a
+                        // nullable-typed pattern
+                        if (IsTrueIdentifier(this.CurrentToken))
+                        {
+                            // If we're in a cast statement/expression arm, then we have:
+                            //
+                            // `switch { case X ? y` or
```

Comment:
This assumption is not correct. The following is [valid code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs7sgDsBnDHWAQIuIXQC8dDFACuMANwcuvAXWw9B2UpLqkllLtwj8tm9cV3EDR2gxR0AsgAoNWmgEo2yo5z4B3bAwwAAtXT29DX2iwAEM+GHEAfnUdEAt04FhYgGsbaLoAXx8iqkKgA):

```cs
        switch(i3) {
            case b ? i1 : i2: break;
        }
```

This comment needs to be corrected, and we should re-examine this approach to ensure that we're not breaking anything.

Reply:
The "why on Earth this is legal" question should have probably been asked N years ago when switch statements were designed) I was sure that conditional expressions in case labels aren't semantically legal and thus there is no difference whether this is parsed as a conditional or a nullable pattern + label statement. There is already a test in the codebase for that case and I initially wrote a comment explaining how these 2 parse strategies both lead to a compiler error thus are identical for our purposes, had to revert that change. And I'm somewhat surprised that only a very few new test cases now are back to bad state after removing this assumption, so I don't think we should do additional work to restore them

Reply:
> The "why on Earth this is legal" question should have probably been asked N years ago when switch statements were designed)

So, this is legal because cases allow for patterns.  And one sort of pattern is the 'constant pattern' (that's how `x is 0` works for example).  So these places allow for any sort of constant expression (including conditional expressions).  Note: for conditional expressions to be constant, their condition itself must be constant.  But that can often normally happen, esp. in generated code, or code referencing constants that are conditionally compiled in.